zerver/tests/test_auth_backends.py

```diff
@@ -328,82 +292,104 @@ def test_remote_user_backend_sso_append_domain(self) -> None:
                 del kwargs['subdomain']
             result = orig_authenticate(backend, *args, **kwargs)
             return result
-        backend.authenticate = patched_authenticate
-        good_kwargs = dict(backend=backend, strategy=backend.strategy,
-                           storage=backend.strategy.storage,
-                           response=token_data_dict,
-                           subdomain='zulip')
-        bad_kwargs = dict(subdomain='acme')
-        with mock.patch('zerver.views.auth.redirect_and_log_into_subdomain',
-                        return_value=user):
-            self.verify_backend(backend,
-                                good_kwargs=good_kwargs,
-                                bad_kwargs=bad_kwargs)
-            bad_kwargs['subdomain'] = "zephyr"
-            self.verify_backend(backend,
-                                good_kwargs=good_kwargs,
-                                bad_kwargs=bad_kwargs)
-        backend.authenticate = orig_authenticate
-        httpretty.disable()
-
-class ResponseMock:
-    def __init__(self, status_code: int, data: Any) -> None:
-        self.status_code = status_code
-        self.data = data
-
-    def json(self) -> str:
-        return self.data
 
-    @property
-    def text(self) -> str:
-        return "Response text"
+        for backend_name in backends_to_test:
+            httpretty.enable(allow_net_connect=False)
+            urls = backends_to_test[backend_name]['urls']   # type: List[Dict[str, Any]]
+            for details in urls:
+                httpretty.register_uri(
+                    details['method'],
+                    details['url'],
+                    status=details['status'],
+                    body=details['body'])
+            backend_class = backends_to_test[backend_name]['backend']
+            backend = backend_class()
+            backend.strategy = DjangoStrategy(storage=BaseDjangoStorage())
+            orig_authenticate = backend_class.authenticate
+            backend.authenticate = patched_authenticate
+            good_kwargs = dict(backend=backend, strategy=backend.strategy,
+                               storage=backend.strategy.storage,
+                               response=token_data_dict,
+                               subdomain='zulip')
+            bad_kwargs = dict(subdomain='acme')
+            with mock.patch('zerver.views.auth.redirect_and_log_into_subdomain',
+                            return_value=user):
+                self.verify_backend(backend,
+                                    good_kwargs=good_kwargs,
+                                    bad_kwargs=bad_kwargs)
+                bad_kwargs['subdomain'] = "zephyr"
+                self.verify_backend(backend,
+                                    good_kwargs=good_kwargs,
+                                    bad_kwargs=bad_kwargs)
+            backend.authenticate = orig_authenticate
+            httpretty.disable()
+            httpretty.reset()
```

Comment:
Most of this test refactoring was a nice improvement that could be done in advance of the migration, so I split this out and merged it as 6b796a1e8c53576b6d06de0f9f9e3ea5100a0ce4.